File: src/repos/resolver.js
/**
 * @module repos/resolver
 * @description Resolves repo directories from config.yaml.
 *
 * Three modes:
 *   - **dir**: one local repo directory
 *   - **parentDir**: parent directory where each subdirectory is a repo
 *   - **clone**: one or more git URLs, cloned on first use
 *
 * Also handles `baseBranch` and `prepareRepo()` to ensure the local
 * repo is on the latest base branch before agents start working.
 */

const fs = require('fs');
const path = require('path');
const { execSync } = require('child_process');
const config = require('../config');
const logger = require('../utils/logger');

/**
 * Get the repo directory for a given component name.
 *
 * - **dir**: always returns config.repo.path (component ignored)
 * - **parentDir**: returns `<path>/<component>` or `<path>` if no component
 * - **clone**: clones URLs into cloneDir, returns the matching repo dir
 *
 * @param {string} [component] - Subdirectory or repo name (for parentDir/clone modes)
 * @returns {string} Absolute path to the repo directory
 */
function getRepoDir(component) {
  const { mode } = config.repo;

  if (mode === 'dir') {
    return config.repo.path;
  }

  if (mode === 'parentDir') {
    if (!component) return config.repo.path;
    return path.join(config.repo.path, component);
  }

  if (mode === 'clone') {
    const cloneDir = config.repo.cloneDir;
    if (!fs.existsSync(cloneDir)) {
      fs.mkdirSync(cloneDir, { recursive: true });
    }

    // Clone all URLs that haven't been cloned yet
    for (const url of config.repo.urls) {
      const repoName = path.basename(url, '.git');
      const targetDir = path.join(cloneDir, repoName);
      if (!fs.existsSync(targetDir)) {
        logger.info(`Cloning ${url} into ${targetDir}`);
        execSync(`git clone ${url} ${targetDir}`, { stdio: 'inherit' });
      }
    }

    // If one URL, return it directly
    if (config.repo.urls.length === 1) {
      const repoName = path.basename(config.repo.urls[0], '.git');
      return path.join(cloneDir, repoName);
    }

    // Multiple URLs — pick by component or return cloneDir
    if (component) {
      return path.join(cloneDir, component);
    }
    return cloneDir;
  }

  throw new Error(`Unknown repo mode: "${mode}". Supported: dir, parentDir, clone`);
}

/**
 * Get the configured base branch.
 * @returns {string} Base branch name (e.g. 'main', 'master', 'develop')
 */
function getBaseBranch() {
  return config.repo.baseBranch;
}

/**
 * Prepare a repo for a new ticket — checkout baseBranch and pull latest.
 * Call this before creating a feature branch.
 *
 * @param {string} repoDir - Absolute path to the repo directory
 */
function prepareRepo(repoDir) {
  const baseBranch = getBaseBranch();
  logger.info(`Preparing repo: checkout ${baseBranch} and pull latest`, { repoDir });
  try {
    execSync(`git checkout ${baseBranch} && git pull`, {
      cwd: repoDir,
      stdio: 'pipe',
    });
  } catch (err) {
    logger.warn(`Failed to prepare repo (may not be a git repo or branch doesn't exist): ${err.message}`);
  }
}

/**
 * List all repo directories.
 * @returns {string[]}
 */
function listRepos() {
  const { mode } = config.repo;

  if (mode === 'dir') {
    return [config.repo.path];
  }

  if (mode === 'parentDir') {
    const parentDir = config.repo.path;
    if (!parentDir || !fs.existsSync(parentDir)) return [];
    return fs.readdirSync(parentDir, { withFileTypes: true })
      .filter(d => d.isDirectory() && !d.name.startsWith('.'))
      .map(d => path.join(parentDir, d.name));
  }

  if (mode === 'clone') {
    const cloneDir = config.repo.cloneDir;
    return config.repo.urls.map(url => {
      const repoName = path.basename(url, '.git');
      return path.join(cloneDir, repoName);
    });
  }

  return [];
}

/**
 * Get the mode from config.
 * @returns {'dir'|'parentDir'|'clone'}
 */
function getMode() {
  return config.repo.mode;
}

module.exports = { getRepoDir, getBaseBranch, prepareRepo, listRepos, getMode };

File: src/repos/resolver.py
"""Repo directory resolver — 3 modes (dir, parentDir, clone) + baseBranch."""

import logging
import subprocess
from pathlib import Path

from src.config import config

logger = logging.getLogger(__name__)


def get_repo_dir(component: str | None = None) -> str:
    """Resolve the repo directory for the given component."""
    mode = config["repo"]["mode"]

    if mode == "dir":
        return config["repo"]["path"]

    if mode == "parentDir":
        base = config["repo"]["path"]
        if not component:
            return base
        return str(Path(base) / component)

    if mode == "clone":
        clone_dir = Path(config["repo"]["clone_dir"])
        clone_dir.mkdir(parents=True, exist_ok=True)

        for url in config["repo"]["urls"]:
            repo_name = Path(url).stem  # strip .git
            target = clone_dir / repo_name
            if not target.exists():
                logger.info(f"Cloning {url} into {target}")
                subprocess.run(["git", "clone", url, str(target)], check=True)

        urls = config["repo"]["urls"]
        if len(urls) == 1:
            return str(clone_dir / Path(urls[0]).stem)

        if component:
            return str(clone_dir / component)
        return str(clone_dir)

    raise ValueError(f"Unknown repo mode: '{mode}'. Supported: dir, parentDir, clone")


def get_base_branch() -> str:
    return config["repo"]["base_branch"]


def prepare_repo(repo_dir: str) -> None:
    """Stash changes, checkout baseBranch, reset to origin."""
    base = get_base_branch()
    logger.info(f"Preparing repo: stash, checkout {base}, reset to origin", extra={"repo_dir": repo_dir})

    def run(cmd):
        try:
            return subprocess.run(cmd, cwd=repo_dir, capture_output=True, text=True).stdout.strip()
        except Exception:
            return None

    result = run(["git", "stash", "--include-untracked"])
    if result and "No local changes" not in result:
        logger.info("Stashed uncommitted changes")

    if run(["git", "checkout", base]) is None:
        logger.warning(f"Failed to checkout {base}")
        return

    run(["git", "fetch", "origin"])

    if run(["git", "reset", "--hard", f"origin/{base}"]) is None:
        run(["git", "pull"])

    logger.info(f"Repo ready on {base}")


def list_repos() -> list[str]:
    mode = config["repo"]["mode"]

    if mode == "dir":
        return [config["repo"]["path"]]

    if mode == "parentDir":
        base = Path(config["repo"]["path"])
        if not base.exists():
            return []
        return [str(d) for d in sorted(base.iterdir()) if d.is_dir() and not d.name.startswith(".")]

    if mode == "clone":
        clone_dir = Path(config["repo"]["clone_dir"])
        return [str(clone_dir / Path(url).stem) for url in config["repo"]["urls"]]

    return []
